main.c

```diff
@@ -3,106 +3,179 @@
+territory in the buffer. This ensures that the messages are always processed FIFO - otherwise, if the 
+producer is pausing before writing more messages, the consumer might pass over the more recent messages and
+read old messages out-of-order. As a consequence, any messages to the right of the current cell that the
+producer has locked are unreachable and will be lost. (Nb - this approach needs amendment when there are 
+multiple producers or consumers.)
+
+Currently this program sends FINAL_MESSAGE + 1 messages through the producer, to the buffer, and to the 
+consumer. The consumer reports an error and kills the program if it detects that it has received messages
+out of order, which indicates a serious failure by the program. At the end, it prints how many messages
+it successfully read.
+
+The PRODUCER_SLOWDOWN and CONSUMER_SLOWDOWN allow you to make producers and consumers randomly lag during 
+operation, to simulate doing expensive operations. Slowing down the consumer leads to far fewer messages
+being recieved, as they are overwritten by the faster producer. Slowing down the producer causes all or 
+nearly all messages to get read by the consumer. The slowdown works by randomly choosing a number x in the 
+range [0, SLOWDOWN] and incrementing an integer from 0 to x in a while loop.
+(nb - it is a very low variance randomness. might be interesting to use an exponential distribution instead
+of uniform)
+
 */
 
-#define BUFFER_SIZE 4
+#define BUFFER_SIZE 1000
+#define FINAL_MESSAGE 500000
+#define PRODUCER_SLOWDOWN 200
+#define CONSUMER_SLOWDOWN 200
 
 struct message {
 	int content;
 	uint32_t locked; 
 };
 
-int finished=0;
+int die_immediately=0;
 struct message buffer[BUFFER_SIZE];
+pthread_t consumer_thread, producer_thread;
 
+void slowdown(int n) {
+	int target = (int) (drand48() * n);
+	int slow = 0;
+	while (slow++ < target) {}
+}
 
-void producer() {
-	int pos = 0;
+/*
+ * Prints the contents of the buffer to stdout.
+ * Construct a string and then print it rather than printing incrementally,
+ * so that the threads don't splatter each other.
+ */
+void printBuffer() {
+	char str[100] = "[";
+	char next[100];
+	int i;
+	for (i=0; i<BUFFER_SIZE; i++) {
+		struct message m = buffer[i];
+		if (m.locked == 1) {
+			strcat(str, "X,");
+		} else {
+			sprintf(next, "%d,", m.content % 100);
+			strcat(str, next);
+		}
+	}
+	strcat(str, "]\n");
+	printf("%s",str);
+}
+
+void *producer(void *_) {
+	int pos = BUFFER_SIZE - 1;
 	buffer[pos].locked = 1;
 	int i;
-	for (i=0; i<10; i++) {
+	for (i=0; i<=FINAL_MESSAGE; i++) {
+		slowdown(PRODUCER_SLOWDOWN);
 		int next = pos;
 		next = (next + 1) % BUFFER_SIZE;
-		// while (1) {
-		// 	if (!OSAtomicTestAndSet(1, &(buffer[next].locked))) {
-		// 		break;
-		// 	}
-		// }
-		printf("at i=%d, %d -> %d\n", next, buffer[next].content, i);
+		while (1) {
+			if (!OSAtomicTestAndSet(1, &(buffer[next].locked))) {
+				break;
+			}
+		}
```

Comment:
This loop indirectly violates the spirit of the contract. If the subscriber has the lock on this slot and takes a very long time, for some reason, then this will busy wait and will prevent the publisher from making progress.